index.js

```diff
@@ -539,15 +609,44 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptionKey) {
     const encryptedVault = this.store.getState().vault;
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    let vault;
+
+    if (this.cacheEncryptionKey) {
+      if (password) {
+        const result = await this.encryptor.decryptWithDetail(
+          password,
+          encryptedVault,
+        );
+        vault = result.vault;
+        this.encryptionSalt = result.salt;
+        this.password = password;
+
+        this.memStore.updateState({ encryptionKey: result.exportedKeyString });
+      } else {
+        const key = await this.encryptor.importKey(encryptionKey);
+        vault = await this.encryptor.decryptWithKey(
+          key,
+          JSON.parse(encryptedVault),
+        );
+
+        this.encryptionSalt = JSON.parse(encryptedVault).salt;
```

Comment:
Nit: We can refactor this to only parse the vault once:

```suggestion
        const parsedEncryptedVault = JSON.parse(encryptedVault);
        vault = await this.encryptor.decryptWithKey(
          key,
          parsedEncryptedVault,
        );

        this.encryptionSalt = parsedEncryptedVault.salt;
```

Probably nicer to rename `parsedEncryptedVault` to `encryptedVault`, and rename the current `encryptedVault` to `serializedEncryptedVault` (or something like that), but that was harder to do in a suggestion.